src/puppetlabs/puppetdb/cli/shovel.clj

```diff
@@ -0,0 +1,378 @@
+            Connection
+            MessageConsumer
+            Queue
+            Message]
+           [java.io File ByteArrayInputStream]
+           [org.apache.activemq.broker BrokerService]
+           [org.apache.activemq.usage SystemUsage MemoryUsage]
+           [org.apache.activemq.pool PooledConnectionFactory]
+           [org.apache.activemq ActiveMQConnectionFactory ScheduledMessage])
+  (:require [clojure.tools.logging :as log]
+            [clojure.java.io :as io]
+            [me.raynes.fs :as fs]
+            [stockpile :as stockpile]
+            [puppetlabs.puppetdb.cheshire :as json]
+            [puppetlabs.i18n.core :as i18n]
+            [puppetlabs.trapperkeeper.config :as config]
+            [puppetlabs.trapperkeeper.logging :as logutils]
+            [puppetlabs.kitchensink.core :as kitchensink]
+            [puppetlabs.puppetdb.config :as conf]
+            [puppetlabs.puppetdb.utils :as utils]
+            [puppetlabs.puppetdb.command.dlo :as dlo]
+            [puppetlabs.kitchensink.core :as ks]
+            [schema.core :as s]))
+
+(def cli-description "Transfer messages from ActiveMQ to Stockpile")
+
+(defn mq-dir [config]
+  (str (io/file (get-in config [:global :vardir]) "mq")))
+
+(defn ^Session create-session [^Connection connection]
+  (.createSession connection false Session/CLIENT_ACKNOWLEDGE))
+
+(defn ^Queue create-queue [^Session session endpoint]
+  (.createQueue session endpoint))
+
+(defn ^Connection create-connection [^ConnectionFactory factory]
+  (doto (.createConnection factory)
+    (.setExceptionListener
+     (reify ExceptionListener
+       (onException [this ex]
+         (log/error ex (i18n/trs "receiver queue connection error")))))
+    (.start)))
+
+(defn extract-headers
+  "Creates a map of custom headers included in `message`, currently only
+  supports String headers."
+  [^Message msg]
+  (into {}
+        (for [k (enumeration-seq (.getPropertyNames msg))]
+          [(keyword k) (.getStringProperty msg k)])))
+
+(defprotocol ConvertMessageBody
+  (convert-message-body ^String [msg]))
+
+(extend-protocol ConvertMessageBody
+  TextMessage
+  (convert-message-body [msg]
+    (-> msg
+        .getText
+        (.getBytes "UTF-8")))
+  BytesMessage
+  (convert-message-body [msg]
+    (let [len (.getBodyLength msg)
+          buf (byte-array len)
+          n (.readBytes msg buf)]
+      (when (not= len n)
+        (throw (Exception. (i18n/trs "Only read {0}/{1} bytes from incoming message" n len))))
+      buf)))
+
+(defn coerce-clj->json-byte-stream
+  "Converts clojure data to JSON serialized bytes on a
+  ByteArrayInputStream. Cheshire will only output to writers and
+  stockpile will only accept input streams, so some conversion needs
+  to be done in this fn"
+  [payload]
+  (let [baos (java.io.ByteArrayOutputStream.)]
+    (with-open [osw (java.io.OutputStreamWriter. baos java.nio.charset.StandardCharsets/UTF_8)]
+      (json/generate-stream payload osw))
+    (.close baos)
```

Comment:
I don't think it's redundant, but it looks like it's not needed https://docs.oracle.com/javase/7/docs/api/java/io/ByteArrayOutputStream.html#close(). I didn't have it in a with-open because I was using the close to ensure all the data was written to the byte array (flushed etc), then I needed to call the toByteArray below. It seems odd not to close a stream, but I can remove that since it's a no-op for a ByteArrayInputStream